File: app/src/main/java/net/devemperor/dictate/rewording/PromptsDatabaseHelper.java
package net.devemperor.dictate.rewording;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.Nullable;

import net.devemperor.dictate.R;

import java.util.ArrayList;
import java.util.List;

public class PromptsDatabaseHelper extends SQLiteOpenHelper {
    private final Context context;

    private static final String DATABASE_NAME = "prompts.db";
    private static final int DATABASE_VERSION = 2;

    public PromptsDatabaseHelper(@Nullable Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        this.context = context;
    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {
        sqLiteDatabase.execSQL("CREATE TABLE PROMPTS (ID INTEGER PRIMARY KEY, POS INTEGER, NAME TEXT, PROMPT TEXT, REQUIRES_SELECTION BOOLEAN, AUTO_APPLY BOOLEAN DEFAULT 0)");

        if (context == null) return;
        ContentValues cv = new ContentValues();
        cv.put("POS", 0);
        cv.put("NAME", context.getString(R.string.dictate_example_prompt_one_name));
        cv.put("PROMPT", context.getString(R.string.dictate_example_prompt_one_prompt));
        cv.put("REQUIRES_SELECTION", 1);
        cv.put("AUTO_APPLY", 0);
        sqLiteDatabase.insert("PROMPTS", null, cv);

        cv = new ContentValues();
        cv.put("POS", 1);
        cv.put("NAME", context.getString(R.string.dictate_example_prompt_two_name));
        cv.put("PROMPT", context.getString(R.string.dictate_example_prompt_two_prompt));
        cv.put("REQUIRES_SELECTION", 1);
        cv.put("AUTO_APPLY", 0);
        sqLiteDatabase.insert("PROMPTS", null, cv);

        cv = new ContentValues();
        cv.put("POS", 2);
        cv.put("NAME", context.getString(R.string.dictate_example_prompt_three_name));
        cv.put("PROMPT", context.getString(R.string.dictate_example_prompt_three_prompt));
        cv.put("REQUIRES_SELECTION", 0);
        cv.put("AUTO_APPLY", 0);
        sqLiteDatabase.insert("PROMPTS", null, cv);

        cv = new ContentValues();
        cv.put("POS", 3);
        cv.put("NAME", context.getString(R.string.dictate_example_prompt_four_name));
        cv.put("PROMPT", context.getString(R.string.dictate_example_prompt_four_prompt));
        cv.put("REQUIRES_SELECTION", 0);
        cv.put("AUTO_APPLY", 0);
        sqLiteDatabase.insert("PROMPTS", null, cv);

        cv = new ContentValues();
        cv.put("POS", 4);
        cv.put("NAME", context.getString(R.string.dictate_example_prompt_five_name));
        cv.put("PROMPT", context.getString(R.string.dictate_example_prompt_five_prompt));
        cv.put("REQUIRES_SELECTION", 0);
        cv.put("AUTO_APPLY", 0);
        sqLiteDatabase.insert("PROMPTS", null, cv);
    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int oldVersion, int newVersion) {
        if (oldVersion < 2) {
            sqLiteDatabase.execSQL("ALTER TABLE PROMPTS ADD COLUMN AUTO_APPLY BOOLEAN DEFAULT 0");
        }
    }

    public int add(PromptModel model) {
        SQLiteDatabase db = getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("POS", model.getPos());
        cv.put("NAME", model.getName());
        cv.put("PROMPT", model.getPrompt());
        cv.put("REQUIRES_SELECTION", model.requiresSelection());
        cv.put("AUTO_APPLY", model.isAutoApply());
        return (int) db.insert("PROMPTS", null, cv);
    }

    public void addAll(List<PromptModel> models) {
        if (models == null || models.isEmpty()) return;
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            for (PromptModel model : models) {
                ContentValues cv = new ContentValues();
                cv.put("POS", model.getPos());
                cv.put("NAME", model.getName());
                cv.put("PROMPT", model.getPrompt());
                cv.put("REQUIRES_SELECTION", model.requiresSelection());
                cv.put("AUTO_APPLY", model.isAutoApply());
                db.insert("PROMPTS", null, cv);
            }
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
            db.close();
        }
    }

    public void replaceAll(List<PromptModel> models) {
        SQLiteDatabase db = getWritableDatabase();
        db.beginTransaction();
        try {
            db.delete("PROMPTS", null, null);
            if (models != null) {
                for (PromptModel model : models) {
                    ContentValues cv = new ContentValues();
                    cv.put("POS", model.getPos());
                    cv.put("NAME", model.getName());
                    cv.put("PROMPT", model.getPrompt());
                    cv.put("REQUIRES_SELECTION", model.requiresSelection());
                    cv.put("AUTO_APPLY", model.isAutoApply());
                    db.insert("PROMPTS", null, cv);
                }
            }
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
            db.close();
        }
    }

    public void update(PromptModel model) {
        SQLiteDatabase db = getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("POS", model.getPos());
        cv.put("NAME", model.getName());
        cv.put("PROMPT", model.getPrompt());
        cv.put("REQUIRES_SELECTION", model.requiresSelection());
        cv.put("AUTO_APPLY", model.isAutoApply());
        db.update("PROMPTS", cv, "ID = " + model.getId(), null);
        db.close();
    }

    public void delete(int id) {
        SQLiteDatabase db = getWritableDatabase();
        db.delete("PROMPTS", "ID = " + id, null);
        db.close();
    }

    public PromptModel get(int id) {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM PROMPTS WHERE ID = " + id, null);
        PromptModel model = null;
        if (cursor.moveToFirst()) {
            model = new PromptModel(cursor.getInt(0), cursor.getInt(1), cursor.getString(2), cursor.getString(3), cursor.getInt(4) == 1, cursor.getInt(5) == 1);
        }
        cursor.close();
        db.close();
        return model;
    }

    public List<PromptModel> getAll() {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM PROMPTS ORDER BY POS", null);

        List<PromptModel> models = new ArrayList<>();
        if (cursor.moveToFirst()) {
            do {
                models.add(new PromptModel(cursor.getInt(0), cursor.getInt(1), cursor.getString(2), cursor.getString(3), cursor.getInt(4) == 1, cursor.getInt(5) == 1));
            } while (cursor.moveToNext());
        }
        cursor.close();
        db.close();
        return models;
    }

    public List<PromptModel> getAllForKeyboard() {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM PROMPTS ORDER BY POS ASC", null);

        List<PromptModel> promptModels = new ArrayList<>();
        if (cursor.moveToFirst()) {
            do {
                promptModels.add(new PromptModel(cursor.getInt(0), cursor.getInt(1), cursor.getString(2), cursor.getString(3), cursor.getInt(4) == 1, cursor.getInt(5) == 1));
            } while (cursor.moveToNext());
        }
        cursor.close();
        db.close();
        List<PromptModel> models = new ArrayList<>(promptModels.size() + 2);
        models.add(new PromptModel(-1, Integer.MIN_VALUE, null, null, false, false));  // Add empty model for instant prompt
        models.add(new PromptModel(-2, Integer.MAX_VALUE, null, null, false, false));  // Add empty model for add button
        models.addAll(promptModels);
        return models;
    }

    public List<Integer> getAutoApplyIds() {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT ID FROM PROMPTS WHERE AUTO_APPLY = 1 ORDER BY POS ASC", null);
        List<Integer> autoApplyIds = new ArrayList<>();
        if (cursor.moveToFirst()) {
            do {
                autoApplyIds.add(cursor.getInt(0));
            } while (cursor.moveToNext());
        }
        cursor.close();
        db.close();
        return autoApplyIds;
    }

    public int count() {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT COUNT(*) FROM PROMPTS", null);
        cursor.moveToFirst();
        int count = cursor.getInt(0);
        cursor.close();
        db.close();
        return count;
    }
}

File: app/src/main/kotlin/dev/patrickgold/florisboard/dictate/data/prompts/PromptsDatabaseHelper.kt
/*
 * Copyright (C) 2026 DevEmperor (Dictate)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

package dev.patrickgold.florisboard.dictate.data.prompts

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import dev.patrickgold.florisboard.R

/**
 * Storage for user-defined rewording prompts.
 *
 * IMPORTANT (data contract – see `docs/COMPATIBILITY.md`):
 * Database name (`prompts.db`), version (`2`) and the exact `PROMPTS` schema are FROZEN so that
 * existing Dictate users keep their prompts after the in-place app update. Do not migrate this to
 * Room without a deliberately tested migration – Room's strict type-affinity validation would
 * reject the legacy `BOOLEAN`/`INTEGER` columns of an existing user database.
 *
 * Lifecycle (issue #138): obtain via [getInstance] and NEVER call `close()` on the returned
 * [SQLiteDatabase]. `SQLiteOpenHelper` hands out one shared, reference-counted database whose
 * connection pool is meant to stay open for the process lifetime. Closing it per query disposed the
 * pool out from under a concurrent, in-flight cursor (overlapping `reload` on the Prompts screen) →
 * `IllegalStateException: connection pool has been closed`. The single process-wide instance is safe
 * to use from multiple threads/coroutines; SQLite's own connection pool serialises access.
 */
class PromptsDatabaseHelper private constructor(
    private val context: Context,
) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(
            "CREATE TABLE PROMPTS (ID INTEGER PRIMARY KEY, POS INTEGER, NAME TEXT, PROMPT TEXT, " +
                "REQUIRES_SELECTION BOOLEAN, AUTO_APPLY BOOLEAN DEFAULT 0)"
        )
        // Seed the example prompts for fresh installs only (existing users skip onCreate). These are
        // the same defaults the legacy Dictate app shipped, resolved from string resources so they are
        // localized (e.g. German) just like the original.
        defaultSeeds().forEachIndexed { index, seed ->
            val cv = ContentValues().apply {
                put("POS", index)
                put("NAME", context.getString(seed.nameRes))
                put("PROMPT", context.getString(seed.promptRes))
                put("REQUIRES_SELECTION", if (seed.requiresSelection) 1 else 0)
                put("AUTO_APPLY", 0)
            }
            db.insert("PROMPTS", null, cv)
        }
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion < 2) {
            db.execSQL("ALTER TABLE PROMPTS ADD COLUMN AUTO_APPLY BOOLEAN DEFAULT 0")
        }
    }

    fun add(model: PromptModel): Int {
        val db = writableDatabase
        val cv = model.toContentValues()
        return db.insert("PROMPTS", null, cv).toInt()
    }

    fun addAll(models: List<PromptModel>?) {
        if (models.isNullOrEmpty()) return
        val db = writableDatabase
        db.beginTransaction()
        try {
            models.forEach { db.insert("PROMPTS", null, it.toContentValues()) }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    fun replaceAll(models: List<PromptModel>?) {
        val db = writableDatabase
        db.beginTransaction()
        try {
            db.delete("PROMPTS", null, null)
            models?.forEach { db.insert("PROMPTS", null, it.toContentValues()) }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    fun update(model: PromptModel) {
        val db = writableDatabase
        db.update("PROMPTS", model.toContentValues(), "ID = ?", arrayOf(model.id.toString()))
    }

    fun delete(id: Int) {
        val db = writableDatabase
        db.delete("PROMPTS", "ID = ?", arrayOf(id.toString()))
    }

    fun get(id: Int): PromptModel? {
        val db = readableDatabase
        return db.rawQuery("SELECT * FROM PROMPTS WHERE ID = ?", arrayOf(id.toString())).use { cursor ->
            if (cursor.moveToFirst()) cursor.toPromptModel() else null
        }
    }

    fun getAll(): List<PromptModel> {
        val db = readableDatabase
        val models = ArrayList<PromptModel>()
        db.rawQuery("SELECT * FROM PROMPTS ORDER BY POS ASC", null).use { cursor ->
            if (cursor.moveToFirst()) {
                do { models.add(cursor.toPromptModel()) } while (cursor.moveToNext())
            }
        }
        return models
    }

    /**
     * Returns all persisted prompts plus the synthetic UI buttons in display order:
     * instant prompt, select-all, [persisted prompts…], add prompt.
     */
    fun getAllForKeyboard(): List<PromptModel> {
        val persisted = getAll()
        val result = ArrayList<PromptModel>(persisted.size + 3)
        result.add(PromptModel(PromptModel.ID_INSTANT_PROMPT, Int.MIN_VALUE, null, null, false, false))
        result.add(PromptModel(PromptModel.ID_SELECT_ALL, Int.MIN_VALUE + 1, null, null, false, false))
        result.addAll(persisted)
        result.add(PromptModel(PromptModel.ID_ADD_PROMPT, Int.MAX_VALUE, null, null, false, false))
        return result
    }

    fun getAutoApplyIds(): List<Int> {
        val db = readableDatabase
        val ids = ArrayList<Int>()
        db.rawQuery("SELECT ID FROM PROMPTS WHERE AUTO_APPLY = 1 ORDER BY POS ASC", null).use { cursor ->
            if (cursor.moveToFirst()) {
                do { ids.add(cursor.getInt(0)) } while (cursor.moveToNext())
            }
        }
        return ids
    }

    fun count(): Int {
        val db = readableDatabase
        return db.rawQuery("SELECT COUNT(*) FROM PROMPTS", null).use { cursor ->
            cursor.moveToFirst()
            cursor.getInt(0)
        }
    }

    private fun PromptModel.toContentValues() = ContentValues().apply {
        put("POS", pos)
        put("NAME", name)
        put("PROMPT", prompt)
        put("REQUIRES_SELECTION", if (requiresSelection) 1 else 0)
        put("AUTO_APPLY", if (autoApply) 1 else 0)
    }

    private fun android.database.Cursor.toPromptModel() = PromptModel(
        id = getInt(getColumnIndexOrThrow("ID")),
        pos = getInt(getColumnIndexOrThrow("POS")),
        name = getString(getColumnIndexOrThrow("NAME")),
        prompt = getString(getColumnIndexOrThrow("PROMPT")),
        requiresSelection = getInt(getColumnIndexOrThrow("REQUIRES_SELECTION")) == 1,
        autoApply = getInt(getColumnIndexOrThrow("AUTO_APPLY")) == 1,
    )

    private data class Seed(val nameRes: Int, val promptRes: Int, val requiresSelection: Boolean)

    companion object {
        const val DATABASE_NAME = "prompts.db"
        const val DATABASE_VERSION = 2

        @Volatile
        private var instance: PromptsDatabaseHelper? = null

        /**
         * The single process-wide helper (issue #138). All callers share one open database instead of
         * each opening — and formerly closing — their own connection, which is what caused overlapping
         * queries to race on a disposed pool. Always built with the application context so it outlives
         * any Activity/Composable that requested it.
         */
        fun getInstance(context: Context): PromptsDatabaseHelper =
            instance ?: synchronized(this) {
                instance ?: PromptsDatabaseHelper(context.applicationContext).also { instance = it }
            }

        // A useful starter set covering all three prompt types: six rewrite-the-selection editors
        // (requiresSelection = true), two free generators and two literal `[snippet]` prompts (both
        // requiresSelection = false). Resolved lazily so the strings pick up the device locale at seed
        // time.
        private fun defaultSeeds() = listOf(
            Seed(R.string.dictate__example_prompt_fix_grammar_name, R.string.dictate__example_prompt_fix_grammar_prompt, true),
            Seed(R.string.dictate__example_prompt_improve_name, R.string.dictate__example_prompt_improve_prompt, true),
            Seed(R.string.dictate__example_prompt_formal_name, R.string.dictate__example_prompt_formal_prompt, true),
            Seed(R.string.dictate__example_prompt_friendly_name, R.string.dictate__example_prompt_friendly_prompt, true),
            Seed(R.string.dictate__example_prompt_shorter_name, R.string.dictate__example_prompt_shorter_prompt, true),
            Seed(R.string.dictate__example_prompt_translate_name, R.string.dictate__example_prompt_translate_prompt, true),
            Seed(R.string.dictate__example_prompt_quote_name, R.string.dictate__example_prompt_quote_prompt, false),
            Seed(R.string.dictate__example_prompt_funfact_name, R.string.dictate__example_prompt_funfact_prompt, false),
            Seed(R.string.dictate__example_prompt_shrug_name, R.string.dictate__example_prompt_shrug_prompt, false),
            Seed(R.string.dictate__example_prompt_signoff_name, R.string.dictate__example_prompt_signoff_prompt, false),
        )
    }
}
